File: SteamLanguageParser/src/main/java/in/dragonbra/steamlanguageparser/parser/SymbolLocator.java
package in.dragonbra.steamlanguageparser.parser;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author lngtr
 * @since 2018-02-15
 */
public class SymbolLocator {
    private static final Pattern IDENTIFIER_REGEX = Pattern.compile("(?<identifier>-?[a-zA-Z0-9_:.]*)");
    private static final Pattern FULL_IDENTIFIER_REGEX = Pattern.compile("(?<class>[a-zA-Z0-9_]*?)::(?<name>[a-zA-Z0-9_]*)");

    private static final Map<String, String> WEAK_TYPES;

    static {
        Map<String, String> weakTypes = new HashMap<>();

        weakTypes.put("byte", "Byte");
        weakTypes.put("short", "Short");
        weakTypes.put("ushort", "Integer");
        weakTypes.put("int", "Integer");
        weakTypes.put("uint", "Long");
        weakTypes.put("long", "Long");
        weakTypes.put("ulong", "Long");

        WEAK_TYPES = Collections.unmodifiableMap(weakTypes);
    }

    private static Node findNode(Node tree, String symbol) {
        return tree.getChildNodes().stream().filter(child -> child.getName().equals(symbol)).findFirst().orElse(null);
    }

    public static Symbol lookupSymbol(Node tree, String identifier, boolean strongOnly) {
        Matcher ident = IDENTIFIER_REGEX.matcher(identifier);

        if (!ident.matches()) {
            throw new IllegalArgumentException("Invalid identifier specified " + identifier);
        }

        if (identifier.contains(".")) {
            String[] split = identifier.split("\\.");

            String val;

            if (split[0].equals("ulong")) {
                switch (split[1]) {
                    case "MaxValue":
                        return new WeakSymbol("0xFFFFFFFFFFFFFFFF");
                    case "MinValue":
                        return new WeakSymbol("0x0000000000000000");
                    default:
                        return new WeakSymbol(identifier);
                }
            }

            switch (split[1]) {
                case "MaxValue":
                    val = "MAX_VALUE";
                    break;
                case "MinValue":
                    val = "MIN_VALUE";
                    break;
                default:
                    return new WeakSymbol(identifier);
            }

            return new WeakSymbol(WEAK_TYPES.get(split[0]) + "." + val);
        } else if (!identifier.contains("::")) {
            Node classNode = findNode(tree, ident.group(0));

            if (classNode == null) {
                if (strongOnly) {
                    throw new IllegalStateException("Invalid weak symbol " + identifier);
                } else {
                    return new WeakSymbol(identifier);
                }
            } else {
                return new StrongSymbol(classNode);
            }
        } else {
            ident = FULL_IDENTIFIER_REGEX.matcher(identifier);

            if (!ident.matches()) {
                throw new IllegalArgumentException("Couldn't parse full identifier " + identifier);
            }

            Node classNode = findNode(tree, ident.group("class"));

            if (classNode == null) {
                throw new IllegalStateException("Invalid class in identifier " + identifier);
            }

            Node propNode = findNode(classNode, ident.group("name"));

            if (propNode == null) {
                throw new IllegalStateException("Invalid property in identifier " + identifier);
            }

            return new StrongSymbol(classNode, propNode);
        }
    }
}

File: buildSrc/src/main/kotlin/in/dragonbra/generators/steamlanguage/parser/symbol/SymbolLocator.kt
package `in`.dragonbra.generators.steamlanguage.parser.symbol

import `in`.dragonbra.generators.steamlanguage.parser.node.Node

class SymbolLocator {
    companion object {
        private val IDENTIFIER_REGEX = "(?<identifier>-?[a-zA-Z0-9_:.]*)".toRegex()
        private val FULL_IDENTIFIER_REGEX = "(?<class>[a-zA-Z0-9_]*?)::(?<name>[a-zA-Z0-9_]*)".toRegex()

        private val WEAK_TYPES: Map<String, String> = mapOf(
            "byte" to "byte",
            "short" to "Short",
            "ushort" to "Integer",
            "int" to "Integer",
            "uint" to "Long",
            "long" to "Long",
            "ulong" to "Long"
        )

        private fun findNode(tree: Node, symbol: String?): Node? =
            tree.childNodes.firstOrNull { child -> child.name == symbol }

        fun lookupSymbol(tree: Node, identifier: String, strongOnly: Boolean): Symbol {
            var ident: MatchResult? = IDENTIFIER_REGEX.matchEntire(identifier)
                ?: throw IllegalArgumentException("Invalid identifier specified $identifier")

            if (identifier.contains(".")) {
                val split = identifier.split(".")

                if (split[0] == "ulong") {
                    when (split[1]) {
                        "MaxValue" -> WeakSymbol("0xFFFFFFFFFFFFFFFF")
                        "MinValue" -> WeakSymbol("0x0000000000000000")
                        else -> WeakSymbol(identifier)
                    }
                }

                val value: String = when (split[1]) {
                    "MaxValue" -> "MAX_VALUE"
                    "MinValue" -> "MIN_VALUE"
                    else -> return WeakSymbol(identifier)
                }

                return WeakSymbol("${WEAK_TYPES[split[0]]}.$value")
            } else if (!identifier.contains("::")) {
                val classNode = findNode(tree, ident?.groupValues?.get(0))

                return if (classNode == null) {
                    if (strongOnly) {
                        throw IllegalStateException("Invalid weak symbol $identifier")
                    }

                    WeakSymbol(identifier)
                } else {
                    StrongSymbol(classNode)
                }
            } else {
                ident = FULL_IDENTIFIER_REGEX.matchEntire(identifier)

                if (ident == null) {
                    throw IllegalArgumentException("Couldn't parse full identifier $identifier")
                }

                val classNode = findNode(tree, ident.groupValues[1])
                    ?: throw IllegalStateException("Invalid class in identifier $identifier")

                val propNode = findNode(classNode, ident.groupValues[2])
                    ?: throw IllegalStateException("Invalid property in identifier $identifier")

                return StrongSymbol(classNode, propNode)
            }
        }
    }
}
